fix: support alerts confirmation page without pending transactions

OGPoyraz opened the pull request:
## Explanation

For more info please visit: https://github.com/MetaMask/MetaMask-planning/issues/404
We should only disable alerts if the confirmation type is `add-ethereum-chain` and without pending transactions.

<!--
Thanks for the pull request. Take a moment to answer these questions so that reviewers have the information they need to properly understand your changes:

* What is the current state of things and why does it need to change?
* What is the solution your changes offer and how does it work?

Are there any issues, Slack conversations, Zendesk issues, user stories, etc. reviewers should consult to understand this pull request better? For instance:

* Fixes #12345
* See: #67890
-->

## Screenshots/Screencaps

This PR has no UI chance.

## Manual Testing Steps

<!--
How should reviewers and QA manually test your changes? For instance:

- Go to this screen
- Do this
- Then do this
-->

## Pre-merge author checklist

- [X] I've clearly explained:
  - [X] What problem this PR is solving
  - [X] How this problem was solved
  - [ ] How reviewers can test my changes
- [ ] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

OGPoyraz committed:
fix: support alerts confirmation page without pending transactions

matthewwalsh0 commented on ui/pages/confirmation/templates/add-ethereum-chain.js:
```diff
@@ -181,7 +181,14 @@ function getState(pendingApproval) {
+  actions,
+  history,
+  _setInputState,
+  unnaprovedTxsCount,
```
Should we put this inside an object such as `state` perhaps so we can group any additional relevant state in future without adding more arguments to the method?

matthewwalsh0 replied:
Should we put this inside an object such as `state` perhaps so we can group any additional relevant state in future without adding more arguments to the method?

matthewwalsh0 commented on ui/pages/confirmation/templates/index.js:
```diff
@@ -127,13 +127,15 @@ function getAttenuatedDispatch(dispatch) {
   dispatch,
   history,
   setInputState,
+  unnaprovedTxsCount,
```
Same here too, so it'd be grouped in both here in the `getTemplateValues` function, but also the `getValues` implementations?

OGPoyraz replied:
Fixed.

matthewwalsh0 commented on ui/pages/confirmation/templates/add-ethereum-chain.js:
```diff
@@ -410,6 +417,7 @@ function getValues(pendingApproval, t, actions, history) {
         ethErrors.provider.userRejectedRequest().serialize(),
       ),
     networkDisplay: !originIsMetaMask,
+    disableAlerts: !(unnaprovedTxsCount > 0),
```
❤️

matthewwalsh0 commented on ui/pages/confirmation/templates/add-ethereum-chain.js:
```diff
@@ -410,6 +417,7 @@ function getValues(pendingApproval, t, actions, history) {
         ethErrors.provider.userRejectedRequest().serialize(),
       ),
     networkDisplay: !originIsMetaMask,
+    disableAlerts: !(unnaprovedTxsCount > 0),
```
Do we have to assume that the `switch-ethereum-chain` would also want this value as it has an alert too and we don't want to change the behaviour?

OGPoyraz replied:
I think this comment is not valid anymore, because we disable alerts in their own context now.

OGPoyraz commented:
> Would this be a great time to add the first unit tests for both `ConfirmationPage` and the templates? Just for the logic we've updated that is.

Added some basic tests just for this alerts.

OGPoyraz commented on ui/pages/confirmation/templates/add-ethereum-chain.js:
```diff
@@ -32,6 +32,7 @@ const UNRECOGNIZED_CHAIN = {
 
 const MISMATCHED_CHAIN_RECOMMENDATION = {
   id: 'MISMATCHED_CHAIN_RECOMMENDATION',
+  severity: SEVERITIES.WARNING,
```
I noticed this was missing while running snapshot tests, adding this will remove some required prop warning.

OGPoyraz committed:
Update snapshot

OGPoyraz committed:
Bump coverage

OGPoyraz commented on ui/pages/confirmation/confirmation.js:
```diff
@@ -91,10 +91,12 @@ const alertStateReducer = produce((state, action) => {
  * the current alert state and function to dismiss an alert by id
  */
-function useAlertState(pendingConfirmation) {
+function useAlertState(pendingConfirmation, { unapprovedTxsCount } = {}) {
```
Destructuring object here intentionally because we don't want to pass entire object to `useEffect` in order to avoid React warnings.

matthewwalsh0 commented on ui/pages/confirmation/confirmation.test.js:
```diff
@@ -0,0 +1,127 @@
+    });
+  });
+
+  describe('switch-ethereum-chain', () => {
```
I appreciate they're technically the same component, but to avoid this file getting too large in future, should we create separate test files for each template?

We could create a test util file with helper functions to avoid unnecessary duplication.

OGPoyraz replied:
Sure that definitely make sense to me.

matthewwalsh0 commented on ui/pages/confirmation/confirmation.test.js:
```diff
@@ -0,0 +1,127 @@
+      });
+    });
+
+    it('should show alert if there are pending txs', async () => {
```
Do we need the inverse test also to be explicit?

OGPoyraz replied:
It's already been asserted in the first snapshot test, do we want to duplicate this test just to check this?

OGPoyraz committed:
separate template tests

OGPoyraz committed:
add snapshots

OGPoyraz committed:
Merge branch 'develop' into fix/support-alerts-confirmationpage-without-pending-transactions

bergarces reviewed (approved):
LGTM
